### DISCUSSION TITLE
Cache misses with same checkout in different build directories

### DISCUSSION THREAD

**Discussion by zjd0112** [ccache/ccache#474] [Support]
### Question ###

I am using ccache to speed up my builds of Contiki OS. But, when I build my application in different directories, I get 0% cache hit. (files in cache are not 0)

- Based on the document, I have set **hash_dir = false**
- One project directory is `/contiki/work-0/examples/hello-world`, and the other project directory is `/contiki/work-0/examples/slip-radio`
- I use Makefile to compile, and change `cc = gcc` to `cc = ccache gcc`
- Based on the log of ccache, I find that the same .o file in two build process are copied to different cache directories, such as '.ccache/4/b/bac91e30c69152e18f7c4171cc434c-24239.o' and '.ccache/2/0/adee5d341253ff58fb90be0955d1df-24241.o'
- The cache size is sufficiently large.

Now, I don't know how to solve this problem. Is there any suggestions?
Thanks in advance for any help!

**Comment by jrosdahl** [comment]
Hi there! It sounds like you should try to set `base_dir` as described on https://ccache.dev/manual/3.7.4.html#_compiling_in_different_directories.

If you still don't get cache hits, there are hints on how to investigate the problem here: https://ccache.dev/manual/3.7.4.html#_cache_debugging

**Comment by zjd0112** [comment]
> Hi there! It sounds like you should try to set `base_dir` as described on https://ccache.dev/manual/3.7.4.html#_compiling_in_different_directories.
> 
> If you still don't get cache hits, there are hints on how to investigate the problem here: https://ccache.dev/manual/3.7.4.html#_cache_debugging

Thank you very much! But, as described on https://ccache.dev/manual/3.7.4.html#_compiling_in_different_directories,  I think if I set **hash_dir = false**, I don't need to set **base_dir**, because the directory info will not be included in the hash value. I wonder that is my understanding correct?

**Comment by jrosdahl** [comment]
> But, as described on https://ccache.dev/manual/3.7.4.html#_compiling_in_different_directories, I think if I set **hash_dir = false**, I don't need to set **base_dir**, because the directory info will not be included in the hash value. I wonder that is my understanding correct?

No, that's not correct. The settings affect different things: **hash_dir** is whether to hash the current directory implicitly while **base_dir** is whether to rewrite absolute command line paths into relative.

**Comment by zjd0112** [comment]
> > But, as described on https://ccache.dev/manual/3.7.4.html#_compiling_in_different_directories, I think if I set **hash_dir = false**, I don't need to set **base_dir**, because the directory info will not be included in the hash value. I wonder that is my understanding correct?
> 
> No, that's not correct. The settings affect different things: **hash_dir** is whether to hash the current directory implicitly while **base_dir** is whether to rewrite absolute command line paths into relative.

Thank you very much! I used the cache debug function just now, it is a pretty good function! I find that the difference in my two compilation processes. Using preprocess model as an example, the only difference of the two "ccache-input-text" file is shown as below.
![test1](https://user-images.githubusercontent.com/24666653/66716054-ad818980-edfc-11e9-97bd-569922f82de1.png)
![test2](https://user-images.githubusercontent.com/24666653/66716062-b1ada700-edfc-11e9-8d04-b455f15fb58e.png)
The difference is shown at line 25. It is because that the output directories in my two compilation processes are different. So, do I have to set the output directory to the same, or is there any other solutions?
Thank you very much!

**Comment by jrosdahl** [comment]
> So, do I have to set the output directory to the same, or is there any other solutions?

Yes, you need to use the same argument to `-MQ`. With different `-MQ` arguments the compilations produce different results (the `.d` file content will be different in the two cases) and ccache thus can't share cache entries. There's an idea in #359 on how to improve this.

**Comment by zjd0112** [comment]
> > So, do I have to set the output directory to the same, or is there any other solutions?
> 
> Yes, you need to use the same argument to `-MQ`. With different `-MQ` arguments the compilations produce different results (the `.d` file content will be different in the two cases) and ccache thus can't share cache entries. There's an idea in #359 on how to improve this.

OK，Thank you very much!
